header/header_errors.go

```diff
@@ -0,0 +1,21 @@
+package header
+
+import (
+	"fmt"
+)
+
+type ErrValidatorHashMismatch struct {
+	Reason error
+}
+
+func (e *ErrValidatorHashMismatch) Error() string {
+	return fmt.Sprintf("validator hash mismatch error: %s", e.Reason.Error())
+}
+
+type ErrLastHeaderHashMismatch struct {
+	Reason error
+}
+
+func (e *ErrLastHeaderHashMismatch) Error() string {
+	return fmt.Sprintf("last header hash mismatch error: %s", e.Reason.Error())
+}
```

Comment:
* Let's make them public only when we need to
* Pls see example here: https://github.com/celestiaorg/go-header/blob/main/verify.go. We don't need to have a reason within an error here. Instead, this can be separate error sentinels with values appended via fmt.Errof

Reply:
Dismiss the first point. I realized that we have test in a separate pkg, so it has to be public.
Please still apply the second comment

Reply:
- They need to be public here since the tests are in a different package `headertest` whereas the header methods are in package `header`.
- I agree we don't need to have a reason specifically but replacing it with an error sentinel could work only when the error string inside `fmt.Errorf` is fixed unless we want to lose the message with actual error hashes:
```
if !bytes.Equal(untrst.LastHeader(), eh.Hash()) {
	return &libhead.VerifyError{
		Reason: &ErrLastHeaderHashMismatch{fmt.Errorf("expected new header to point to last header hash (%X), but got %X)",
			eh.Hash(),
			untrst.LastHeader(),
		)},
	}
}
```

That's because detecting the error is of that certain type in using an error type like in:

```
switch reason.(type) {
	case *header.ErrValidatorHashMismatch:
		assert.Equal(t, &header.ErrValidatorHashMismatch{}, testReason)
}
```

would have to be switched to comparing the error message instead which in the above case would have to do pattern matching instead of simply being able to look at the error struct type.

cc @MSevey since he originally suggested opening an issue about this: https://github.com/rollkit/rollkit/pull/1026#discussion_r1248284884

Reply:
>     but replacing it with an error sentinel could work only when the error string inside fmt.Errorf is fixed unless we want to lose the message with actual error hashes:

See https://pkg.go.dev/errors#Is. When you do `fmt.Errorf("%w", ErrValidatorHashMismatch)`, you can still compare against `ErrValidatorHashMismatch` by using `errors.Is` or `assert.ErrorIs`

Reply:
Yes, I understand. Would you not lose actual error hashes which are formatted in the error string then? 

Currently, it looks like:
```
&ErrLastHeaderHashMismatch{fmt.Errorf("expected new header to point to last header hash (%X), but got %X)",
	eh.Hash(),
	untrst.LastHeader(),
)}
```

I know this works well for constant errors like here: https://github.com/celestiaorg/go-header/blob/2fd3b1bb637db76d5bbecf0731ec100dba6a7c02/verify.go#L88 or https://github.com/rollkit/rollkit/pull/1165/files#diff-d178d1a180ac9d7016ec909f5f2c8c4f0a2fc1bde0c72cc60a28b2788ef4183eR10

Reply:
Used error sentinels and wrapped errors with them instead. Feel free to take a look again 😄

Reply:
@Manav-Aggarwal, wait, %w is just an example, you do this:

`fmt.Errorf("%w: expected old header next validators (%X) to match those from new header (%X)", err, values...)`